### ISSUE TITLE
Tests Needs

### ISSUE THREAD

**Issue by aligoren** [apache/kibble-1#27]
Hi.

I know there is not many contributors here.

But even like that, I think we should write tests.

Before then I think some modules should be rewritten. I know this isn’t easy.

Every developer in this project has its work like me.

Maybe we think a new structure with Flask or an async framework like Bocadillo.

Maybe we talk about the documentation in a different issue.

**Comment by Humbedooh** [comment]
Tests would be awesome. I'm not sure where to start - perhaps with the server API? I know there have been some shortcuts made to speed up processes in the OpenAPI department, that should probably get fixed :) (such as response definitions left very open)

**Comment by aligoren** [comment]
Hehe :) yes. I’ll **read** whole codebase to understand how it works.

Because there are also no docstrings.

I’ll try to find friends to make this project much betters.

**Comment by michalslowikowski00** [comment]
Hey guys.
Maybe pre-commit will be a good idea to starts with? 
I can do it, a least try to do it. 
https://pre-commit.com/

**Comment by turbaszek** [comment]
+1 for introducing pre-commit. It will allow us to assert code quality as a first step and then we can add tests that can also be run using this tool (+ pytest). 

If someone is not familiar with this framework - it's a simple tool that allows users to share and install git hooks that are run before commit. In this way the code quality (and more) is checked even before making a contribution. Of course - it's not necessary but it helps a lot. We use this in Apache Airflow to automatically verify multiple things such as code style, language style (no black and white lists and so on) and formatting of documentation 👍 

Probably it would be good to consider using some CI to avoid regression in the future. WDYT @Humbedooh @sharanf @michalslowikowski00 ?

**Comment by michalslowikowski00** [comment]
@turbaszek yes, absolutely, +1 for the CI. I thought that the Circle CI is implemented here. But it is not.

**Comment by sharanf** [comment]
@michalslowikowski00 thanks for the suggestion. And it sounds like a good practical idea - I like it. Let me check any other community feedback.

**Comment by michalslowikowski00** [comment]
@sharanf should we introduce some proposals via the dev list?

**Comment by sharanf** [comment]
@michalslowikowski00 yes for new stuff, it is best to see if others have an opinion and you will see that I have already started this discussion there :-)

**Comment by michalslowikowski00** [comment]
I subscribed dev list. :)

**Comment by turbaszek** [comment]
Ok, so we've got CI up and running. Now it's time to think about testing 😄 

From discussions in #38 and #48 it seems that it would be wise to rethink the backend of API (+1 for using a framework as @aligoren suggested). But before we decide on something I think it would be a good step to add test coverage for existing API so we are sure that the UI will still work.

Currently, there's 61 API endpoints defined in openapi spec, each can be used with ~2 methods, so we need about 120 test. That's is not that much if we make those test generic and easy to write.
